File: src/exception_ffi.erl
-module(exception_ffi).

-export([rescue/1, defer/2, on_crash/2]).

rescue(F) ->
    try {ok, F()}
    catch
      error:Term -> {error, {errored, Term}};
      throw:Term -> {error, {thrown, Term}};
      exit:Term -> {error, {exited, Term}}
    end.

defer(Cleanup, Body) ->
    try Body()
    after Cleanup()
    end.

on_crash(Cleanup, Body) ->
    try Body()
    catch A:B:C ->
        Cleanup(), 
        erlang:raise(A,B,C)
    end.

File: src/exception_ffi.mjs
import { Ok, Error as GError } from "./gleam.mjs";
import { Errored, Thrown } from "./exception.mjs";

export function on_crash(cleanup, body) {
  try {
    return body();
  } catch (e) {
    cleanup();
    throw e;
  }
}

export function rescue(f) {
  try {
    return new Ok(f());
  } catch (e) {
    if (e instanceof Error) {
      return new GError(new Errored(e));
    } else {
      return new GError(new Thrown(e));
    }
  }
}

export function defer(cleanup, body) {
  try {
    return body();
  } finally {
    cleanup();
  }
}
